WIP - BPF Native NodePort support

tgraf opened the pull request:
This implements NodePort support for both IPv4 and IPv6.

- Forward translation from nodeIP:nodePort => podIP:port happens in BPF on all nodes
- Original source IP of client is preserved
- Reverse translation happens on node where endpoint relies (direct-server-return)
- Packets will look like they come from reserved:world from a policy perspective

Fixes: #869

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/1259)
<!-- Reviewable:end -->

ianvernon commented on daemon/k8s_watcher.go:
```diff
@@ -408,9 +409,19 @@ func (d *Daemon) serviceAddFn(obj interface{}) {
-		break
+	// LoadBalancer and NodePort make the service available on the nodePort
+	// on each node. This creates a service with both the IPv4 and IPv6
+	// address of the host as the frontend.
```
The commentary throughout this patch is very clear; this is a model for quality for what should be done throughout the project!

tgraf committed:
agent: Restrict host=>endpoint masquerade rule to local addresses

The masquerade rule ensures that all traffic from host to endpoints is
masquerade behind an address that is known to Cilium. This should not
apply to traffic that has been loadbalanced to cilium_host from
bpf_netdev.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
agent: Synchronize all service IDs as reverse NAT mappings to datapath

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
bpf_netdev: Add connection tracking based reverse NAT

- Adds connection tracking to connections which have been loadbalanced
  and stores the reverse nat ID in the entry. The global CT is being
  used for this purpose.
- Removes the shortcut in bpf_lxc which avoided the tail call to the
  bpf_netdev policy program when policy enforcement was disabled. This
  tail call is now always required because of eventual reverse NAT in
  the bpf_netdev policy program.
- We now need to disable rp_filter on cilium_host and cilium_net
- Removes the calls_map of bpf_netdev programs before inserting the
  program to alllow to upgrade the map size.
- The loadbalancing and local delivery code of bpf_netdev is split into
  two individual tail called programs to avoid hitting verifier
  complexity limits.
- The reverse NAT ID is carried in the tunnel id in overlay mode for
  IPv4 and in the destination address for IPv6.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
k8s: Automatically match IPv4/IPv6 frontend and backend addresses

THe existing behaviour was to error out if one endpoint address did not
match the IP address family of the frontend. Instead, match any
available endpoint address representing the backend with the matching
frontend address family.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
bpf: Leave CT tuple untouched

This restructures the conntrack code to leave the tuple argument intact
by creating a new copy for the reverse tuple. Leads to slightly
increased stack usage but simplifies the code a lot.

Fixes: #1289

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
k8s: Add NodePort support

Fixes: #869

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
endpoint: Set accept_local=1 and send_redirects=0 on veth

NodePort loadbalancers will cause packets to be emitted to the veth with
the source address set to the host's address. This would trigger the
martian source check in the Linux routing layer. Disable the check when
we are coming from Cilium managed veths.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
agent: Add --direct-server-return flag to enable DSR

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
bpf: Add virtual NAT box

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
bpf: Do not clear policy skip flag in ipv6_policy()

This caused egress CIDR policy to drop even when it was allowed

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
daemon: Enable L3/L4 load-balancer via node_config.h

Load-balancer was only enabled via lxc_config.h so far.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
bpf: Enable debug messages for services in bpf_netdev

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
agent: Delete feeder rules before deleting custom chains

All feeder rules must be gone before we can flush and delete the custom
chains.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
agent: No longer allow to change ConntrackLocal on agent level

Related-to: #1404

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
bpf: New bpf_host_pre program to preserve identity

bpf_netdev for packets from the hosts has been moved to the egress side
of cilium_host. This allowed to access the skb->mark from the local
task. It makes it impossible to feed into the NAT box though. Therefore,
Introduce a small bpf_host_pre.c which runs at egress and converts the
identity from skb->mark to skb->tc_index which is preserved across the
veth transmition path. This allows to move the main bpf_netdev back to
ingress where it can pass packets back into the stack.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
bpf: Use setup_veth for host pair veth

Relaxes constraints with send_redirects=0 and accept_local=1

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
bpf: Protect undefined SECLABEL

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
bpf: Create ICMPv6 conntrack related entries with correct protocol

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
bpf: Use MAC of cilium_net when passing to stack from host->cilium

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
bpf: Mark as PACKET_HOST when going to NAT box

Also rewrites destination MAC in bpf_host_pre to peer veth

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
bpf: Reverse translate wire packets in bpf_netdev

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
bpf: Clear skb->mark and skb->tc_index after receiving metadata

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
bpf: Trace packets as they go back into the host via bpf_netdev

Signed-off-by: Thomas Graf <thomas@cilium.io>
